File: src/introgression.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! PhyloNet-HMM for introgression detection (Paper 018).
//!
//! Liu et al. (2015) "Interspecific Introgressive Origin of Genomic
//! Diversity in the House Mouse" PNAS 112:196-201.
//!
//! Port of `control/introgression/introgression.py`.
//! Uses [`crate::hmm::Hmm`] for forward, backward, and Viterbi.
//!
//! ## `BarraCUDA` connection
//!
//! - Forward/backward/Viterbi: delegates to [`crate::hmm`] which maps to
//!   `barracuda::staging::StatefulPipeline` and `barracuda::ops::logsumexp`
//! - Likelihood ratio test: scalar reduction (chi-squared from `barracuda::stats`)
//! - Gene tree topology switching: discrete HMM state space (GPU via `hmm_forward_log.wgsl`)

#![expect(
    clippy::cast_precision_loss,
    reason = "domain-specific numeric patterns"
)]

use crate::hmm::Hmm;
use crate::rng::Rng;

/// Gene tree topology: concordant, introgression-like, or other.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(usize)]
pub enum GeneTreeTopology {
    /// ((B,C),A) — concordant with species tree (ILS).
    Concordant = 0,
    /// ((A,B),C) — introgression-like (A→B gene flow).
    IntrogressionLike = 1,
    /// ((A,C),B) — other.
    Other = 2,
}

impl GeneTreeTopology {
    /// Observation index for HMM emission matrix.
    #[must_use]
    pub const fn to_obs(&self) -> usize {
        *self as usize
    }
}

impl From<usize> for GeneTreeTopology {
    fn from(i: usize) -> Self {
        match i {
            0 => Self::Concordant,
            1 => Self::IntrogressionLike,
            _ => Self::Other,
        }
    }
}

/// PhyloNet-HMM parameters matching Python baseline.
/// 2 states: `ILS_only` (0), Introgression (1). 3 observations.
#[must_use]
pub fn phylonet_hmm() -> Hmm {
    let transition = vec![vec![0.98, 0.02], vec![0.05, 0.95]];
    let emission = vec![vec![0.70, 0.06, 0.24], vec![0.15, 0.75, 0.10]];
    let initial = vec![0.70, 0.30];
    Hmm::new(transition, emission, initial)
}

/// ILS-only model: single state, emission from ILS row only.
#[must_use]
pub fn ils_only_hmm() -> Hmm {
    let emission = vec![vec![0.70, 0.06, 0.24]];
    let transition = vec![vec![1.0]];
    let initial = vec![1.0];
    Hmm::new(transition, emission, initial)
}

/// HMM that stays in ILS forever (for generating no-introgression data).
#[must_use]
fn ils_only_generating_hmm() -> Hmm {
    let emission = vec![vec![0.70, 0.06, 0.24], vec![0.70, 0.06, 0.24]];
    let transition = vec![vec![1.0, 0.0], vec![0.0, 1.0]];
    let initial = vec![1.0, 0.0];
    Hmm::new(transition, emission, initial)
}

/// Generate synthetic loci with no introgression (for FPR test).
#[must_use]
pub fn generate_ils_only_loci(n_loci: usize, rng: &mut Rng) -> (Vec<usize>, Vec<usize>) {
    ils_only_generating_hmm().generate_sequence(n_loci, rng)
}

/// Generate synthetic gene tree observations from PhyloNet-HMM.
#[must_use]
pub fn generate_synthetic_loci(
    n_loci: usize,
    hmm: &Hmm,
    rng: &mut Rng,
) -> (Vec<usize>, Vec<usize>) {
    hmm.generate_sequence(n_loci, rng)
}

/// Log-likelihood ratio: 2 * (log `L_introg` - log `L_ils_only`).
/// Positive = introgression model preferred.
#[must_use]
pub fn log_likelihood_ratio(log_lik_introg: f64, log_lik_ils_only: f64) -> f64 {
    2.0 * (log_lik_introg - log_lik_ils_only)
}

/// Detect introgression regions via Viterbi decoding.
/// Returns (path, `log_prob`). `path[i]` = 0 for ILS, 1 for Introgression.
#[must_use]
pub fn detect_introgression(hmm: &Hmm, observations: &[usize]) -> (Vec<usize>, f64) {
    hmm.viterbi(observations)
}

/// Fraction of loci in introgression state.
#[must_use]
pub fn introgression_fraction(path: &[usize]) -> f64 {
    if path.is_empty() {
        return 0.0;
    }
    path.iter().filter(|&&s| s == 1).count() as f64 / path.len() as f64
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tolerances;

    #[test]
    fn topology_to_obs() {
        assert_eq!(GeneTreeTopology::Concordant.to_obs(), 0);
        assert_eq!(GeneTreeTopology::IntrogressionLike.to_obs(), 1);
        assert_eq!(GeneTreeTopology::Other.to_obs(), 2);
    }

    #[test]
    fn phylonet_hmm_forward_finite() {
        let hmm = phylonet_hmm();
        let mut rng = Rng::new(42);
        let (_, obs) = generate_synthetic_loci(500, &hmm, &mut rng);
        let (_, log_lik) = hmm.forward(&obs);
        assert!(log_lik.is_finite());
    }

    #[test]
    fn viterbi_accuracy_above_chance() {
        let hmm = phylonet_hmm();
        let mut rng = Rng::new(42);
        let (true_states, obs) = generate_synthetic_loci(500, &hmm, &mut rng);
        let (path, _) = hmm.viterbi(&obs);
        let acc = path
            .iter()
            .zip(true_states.iter())
            .filter(|(a, b)| a == b)
            .count() as f64
            / path.len() as f64;
        assert!(acc > 0.55);
    }

    #[test]
    fn ils_only_hmm_valid() {
        let hmm = ils_only_hmm();
        assert_eq!(hmm.num_states(), 1);
        let mut rng = Rng::new(42);
        let (states, obs) = generate_ils_only_loci(100, &mut rng);
        assert_eq!(states.len(), 100);
        assert_eq!(obs.len(), 100);
        let (_, loglik) = hmm.forward(&obs);
        assert!(loglik.is_finite());
    }

    #[test]
    fn log_likelihood_ratio_sign() {
        assert!(log_likelihood_ratio(-100.0, -150.0) > 0.0);
        assert!((log_likelihood_ratio(-100.0, -100.0)).abs() < tolerances::CROSS_LANGUAGE);
    }

    #[test]
    fn detect_introgression_returns_path() {
        let hmm = phylonet_hmm();
        let mut rng = Rng::new(42);
        let (_, obs) = generate_synthetic_loci(200, &hmm, &mut rng);
        let (path, loglik) = detect_introgression(&hmm, &obs);
        assert_eq!(path.len(), 200);
        assert!(loglik.is_finite());
    }

    #[test]
    fn introgression_fraction_bounded() {
        let path = vec![0, 0, 1, 1, 0, 1, 0, 0, 1, 1];
        let frac = introgression_fraction(&path);
        assert!((0.0..=1.0).contains(&frac));
        assert!((frac - 0.5).abs() < tolerances::CROSS_LANGUAGE);
    }
}

File: control/introgression/introgression.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Provenance: see src/provenance/experiments.rs — INTROGRESSION_PROVENANCE

#!/usr/bin/env python3
"""
neuralSpring Paper 018 — PhyloNet-HMM for Introgression Detection

Reproduces the computational core from:
  Liu et al. (2015)
  "Interspecific Introgressive Origin of Genomic Diversity in the House Mouse"
  PNAS 112:196-201.

Core thesis: Introgression (gene flow between species after hybridization)
can be detected using statistical methods on genomic data. Uses PhyloNet-HMM
to distinguish introgression from incomplete lineage sorting (ILS).

Species tree: ((B,C),A) — B and C are sister taxa, A is outgroup.
In introgression regions, gene flow A→B yields gene tree ((A,B),C).
Maps to ecoPrimals: introgression = transfer learning between species.

BarraCUDA connection:
  - Forward/backward: gemm_f64.wgsl (transition × emission × state)
  - Viterbi: reduce_max.wgsl + argmax
  - Same HMM primitives as Paper 016 (hmm_phylo).
"""

import sys

import numpy as np

# ---------------------------------------------------------------------------
# Gene tree topology encoding
# ---------------------------------------------------------------------------
# 0: ((B,C),A) — concordant with species tree (ILS)
# 1: ((A,B),C) — introgression-like (A→B gene flow)
# 2: ((A,C),B) — other (ILS or noise)
CONCORDANT = 0
INTROG_LIKE = 1
OTHER = 2


def build_phylonet_hmm(
    ils_concordant_prob: float = 0.7,
    introg_concordant_prob: float = 0.15,
    introg_self_transition: float = 0.95,
    ils_self_transition: float = 0.98,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Build PhyloNet-HMM: 2 states (ILS_only, Introgression), 3 observations.

    ILS_only: gene trees mostly concordant; occasional ILS yields other topologies.
    Introgression: gene trees mostly ((A,B),C); some concordant from ILS.

    Returns (transition, emission, initial).
    """
    # Emission: P(obs | state). Rows = states, cols = [concordant, introg-like, other]
    # ILS_only: high concordant, low introg-like
    # Introgression: high introg-like, low concordant
    p_ils_c = ils_concordant_prob
    p_ils_i = (1.0 - p_ils_c) * 0.2  # low introg-like under ILS
    p_ils_o = 1.0 - p_ils_c - p_ils_i

    p_int_c = introg_concordant_prob
    p_int_i = 0.75  # high introg-like under introgression
    p_int_o = 1.0 - p_int_c - p_int_i

    emission = np.array(
        [
            [p_ils_c, p_ils_i, p_ils_o],
            [p_int_c, p_int_i, p_int_o],
        ],
        dtype=np.float64,
    )

    # Transition: persistence of introgression blocks (high self-transition)
    a_ils = ils_self_transition
    a_int = introg_self_transition
    transition = np.array(
        [
            [a_ils, 1.0 - a_ils],
            [1.0 - a_int, a_int],
        ],
        dtype=np.float64,
    )

    # Initial: ~30% introgression to match target fraction
    initial = np.array([0.70, 0.30], dtype=np.float64)

    return transition, emission, initial


def generate_synthetic_loci(
    n_loci: int,
    transition: np.ndarray,
    emission: np.ndarray,
    initial: np.ndarray,
    seed: int = 42,
) -> tuple[np.ndarray, np.ndarray]:
    """Generate synthetic gene tree topologies from PhyloNet-HMM.

    Returns (true_states, observations). States: 0=ILS, 1=Introgression.
    With initial=[0.7,0.3] and tuned transitions, yields ~30% introgression.
    """
    rng = np.random.default_rng(seed)

    states = np.zeros(n_loci, dtype=int)
    states[0] = rng.choice(2, p=initial)
    for t in range(1, n_loci):
        states[t] = rng.choice(2, p=transition[states[t - 1]])

    observations = np.zeros(n_loci, dtype=int)
    for t in range(n_loci):
        observations[t] = rng.choice(3, p=emission[states[t]])

    return states, observations


class PhyloNetHMM:
    """PhyloNet-HMM for introgression detection. Wraps standard HMM."""

    def __init__(
        self,
        transition: np.ndarray,
        emission: np.ndarray,
        initial: np.ndarray,
    ):
        self.A = np.array(transition, dtype=np.float64)
        self.B = np.array(emission, dtype=np.float64)
        self.pi = np.array(initial, dtype=np.float64)
        self.N = self.A.shape[0]
        self.M = self.B.shape[1]

    def forward(self, observations: np.ndarray) -> tuple[np.ndarray, float]:
        """Forward algorithm: α_t(i) and log-likelihood."""
        T = len(observations)
        alpha = np.zeros((T, self.N))
        scales = np.zeros(T)

        obs0 = min(int(observations[0]), self.M - 1)
        alpha[0] = self.pi * self.B[:, obs0]
        scales[0] = alpha[0].sum()
        alpha[0] /= scales[0]

        for t in range(1, T):
            obt = min(int(observations[t]), self.M - 1)
            alpha[t] = (alpha[t - 1] @ self.A) * self.B[:, obt]
            scales[t] = alpha[t].sum()
            if scales[t] > 0:
                alpha[t] /= scales[t]

        log_lik = float(np.sum(np.log(scales + 1e-300)))
        return alpha, log_lik

    def backward(self, observations: np.ndarray, scales: np.ndarray) -> np.ndarray:
        """Backward algorithm: β_t(i)."""
        T = len(observations)
        beta = np.zeros((T, self.N))
        beta[-1] = 1.0

        for t in range(T - 2, -1, -1):
            obt = min(int(observations[t + 1]), self.M - 1)
            beta[t] = self.A @ (self.B[:, obt] * beta[t + 1])
            if scales[t + 1] > 0:
                beta[t] /= scales[t + 1]

        return beta

    def viterbi(self, observations: np.ndarray) -> tuple[np.ndarray, float]:
        """Viterbi: most likely state sequence."""
        T = len(observations)
        log_A = np.log(self.A + 1e-300)
        log_B = np.log(self.B + 1e-300)
        log_pi = np.log(self.pi + 1e-300)

        delta = np.zeros((T, self.N))
        psi = np.zeros((T, self.N), dtype=int)

        obs0 = min(int(observations[0]), self.M - 1)
        delta[0] = log_pi + log_B[:, obs0]

        for t in range(1, T):
            obt = min(int(observations[t]), self.M - 1)
            for j in range(self.N):
                candidates = delta[t - 1] + log_A[:, j]
                psi[t, j] = np.argmax(candidates)
                delta[t, j] = candidates[psi[t, j]] + log_B[j, obt]

        path = np.zeros(T, dtype=int)
        path[-1] = np.argmax(delta[-1])
        log_prob = float(delta[-1, path[-1]])

        for t in range(T - 2, -1, -1):
            path[t] = psi[t + 1, path[t + 1]]

        return path, log_prob

    def posterior(self, observations: np.ndarray) -> np.ndarray:
        """Posterior P(s_t=i | O) via forward-backward."""
        alpha, _ = self.forward(observations)
        T = len(observations)
        scales = np.zeros(T)
        a = np.zeros((T, self.N))
        obs0 = min(int(observations[0]), self.M - 1)
        a[0] = self.pi * self.B[:, obs0]
        scales[0] = a[0].sum()
        a[0] /= scales[0]
        for t in range(1, T):
            obt = min(int(observations[t]), self.M - 1)
            a[t] = (a[t - 1] @ self.A) * self.B[:, obt]
            scales[t] = a[t].sum()
            if scales[t] > 0:
                a[t] /= scales[t]

        beta = self.backward(observations, scales)
        gamma = alpha * beta
        row_sums = gamma.sum(axis=1, keepdims=True)
        row_sums[row_sums == 0] = 1
        gamma /= row_sums
        return gamma


def build_ils_only_model(
    transition: np.ndarray,
    emission: np.ndarray,
    initial: np.ndarray,
) -> PhyloNetHMM:
    """ILS-only model: force state 0 (ILS) for all loci. Single-state HMM."""
    # Emission: only ILS row; transition trivial; initial [1]
    em_ils = emission[0:1, :]
    trans_ils = np.array([[1.0]])
    init_ils = np.array([1.0])
    return PhyloNetHMM(trans_ils, em_ils, init_ils)


def main() -> int:
    """Validate PhyloNet-HMM introgression detection.

    Provenance
    ----------
    Paper: Liu et al. (2015) PNAS 112:196-201.
    Model: PhyloNet-HMM for introgression vs ILS.
    Validation: forward/Viterbi, LRT, posterior, detection accuracy.
    """
    total_passed = 0
    total_failed = 0
    seed = 42
    n_loci = 500

    print("=" * 72)
    print("neuralSpring Paper 018: PhyloNet-HMM Introgression Detection")
    print("  Liu et al. (2015) PNAS 112:196-201")
    print("=" * 72)

    trans, emission, initial = build_phylonet_hmm()
    model = PhyloNetHMM(trans, emission, initial)

    true_states, obs = generate_synthetic_loci(n_loci, trans, emission, initial, seed=seed)
    true_introg_frac = np.mean(true_states == 1)

    # ------------------------------------------------------------------
    # 1. Forward algorithm produces finite log-likelihood
    # ------------------------------------------------------------------
    print("\n--- Check 1: Forward log-likelihood finite ---")
    _, log_lik = model.forward(obs)
    if np.isfinite(log_lik) and log_lik < 0:
        print(f"  [PASS] Forward log-likelihood: {log_lik:.4f}")
        total_passed += 1
    else:
        print(f"  [FAIL] Forward log-likelihood: {log_lik}")
        total_failed += 1

    # ------------------------------------------------------------------
    # 2. Viterbi path identifies introgression with accuracy > random
    # ------------------------------------------------------------------
    print("\n--- Check 2: Viterbi accuracy > random ---")
    path, viterbi_prob = model.viterbi(obs)
    accuracy = np.mean(path == true_states)
    chance = 0.5
    if accuracy > chance + 0.05:
        print(f"  [PASS] Viterbi accuracy ({accuracy:.4f}) > chance+0.05 ({chance + 0.05:.4f})")
        total_passed += 1
    else:
        print(f"  [FAIL] Viterbi accuracy ({accuracy:.4f}) not above chance")
        total_failed += 1

    # ------------------------------------------------------------------
    # 3. Introgression model has higher log-likelihood than ILS-only
    # ------------------------------------------------------------------
    print("\n--- Check 3: Introgression model preferred over ILS-only ---")
    ils_model = build_ils_only_model(trans, emission, initial)
    _, log_lik_ils = ils_model.forward(obs)
    lr = 2.0 * (log_lik - log_lik_ils)
    if lr > 0:
        print(f"  [PASS] LRT: introg model log-lik ({log_lik:.2f}) > ILS-only ({log_lik_ils:.2f})")
        total_passed += 1
    else:
        print(f"  [FAIL] ILS-only has higher log-lik (LR={lr:.2f})")
        total_failed += 1

    # ------------------------------------------------------------------
    # 4. Posterior probabilities sum to 1 per locus
    # ------------------------------------------------------------------
    print("\n--- Check 4: Posterior sums to 1 per locus ---")
    gamma = model.posterior(obs)
    post_sums = gamma.sum(axis=1)
    if np.allclose(post_sums, 1.0, atol=1e-8):
        print("  [PASS] Posterior sums to 1 at each locus")
        total_passed += 1
    else:
        max_dev = np.max(np.abs(post_sums - 1.0))
        print(f"  [FAIL] Posterior deviation: {max_dev:.2e}")
        total_failed += 1

    # ------------------------------------------------------------------
    # 5. Detected introgression fraction near true (within tolerance)
    # ------------------------------------------------------------------
    print("\n--- Check 5: Detected fraction near true ---")
    detected_frac = np.mean(path == 1)
    tol_frac = 0.15
    if abs(detected_frac - true_introg_frac) <= tol_frac:
        print(
            f"  [PASS] Detected {detected_frac:.3f} vs true {true_introg_frac:.3f} (tol={tol_frac})"
        )
        total_passed += 1
    else:
        print(f"  [FAIL] Detected {detected_frac:.3f} vs true {true_introg_frac:.3f}")
        total_failed += 1

    # ------------------------------------------------------------------
    # 6. False positive rate low when no introgression
    # ------------------------------------------------------------------
    print("\n--- Check 6: False positive rate when no introgression ---")
    # ILS-only: initial [1,0], transition [[1,0],[0,1]] keeps in ILS
    trans_pure_ils = np.array([[1.0, 0.0], [0.0, 1.0]])
    true_ils_only, obs_ils_only = generate_synthetic_loci(
        n_loci, trans_pure_ils, emission, np.array([1.0, 0.0]), seed=seed + 1
    )
    # Use full model (not ils_model_only) to test FPR when data is ILS-only
    path_no_introg, _ = model.viterbi(obs_ils_only)
    fp_rate = np.mean(path_no_introg == 1)
    if fp_rate < 0.25:
        print(f"  [PASS] FPR when no introgression: {fp_rate:.3f} < 0.25")
        total_passed += 1
    else:
        print(f"  [FAIL] FPR too high: {fp_rate:.3f}")
        total_failed += 1

    # ------------------------------------------------------------------
    # 7. Gene tree topologies follow expected frequencies
    # ------------------------------------------------------------------
    print("\n--- Check 7: Gene tree topology frequencies ---")
    obs_counts = np.bincount(obs, minlength=3)
    obs_frac = obs_counts / n_loci
    # Under ILS+introgression, expect mix: concordant dominant, some introg-like
    has_concordant = obs_frac[CONCORDANT] > 0.2
    has_introg_like = obs_frac[INTROG_LIKE] > 0.05
    if has_concordant and has_introg_like:
        print(
            f"  [PASS] Concordant: {obs_frac[CONCORDANT]:.3f}, "
            f"Introg-like: {obs_frac[INTROG_LIKE]:.3f}"
        )
        total_passed += 1
    else:
        print(f"  [FAIL] Topology fractions: {obs_frac}")
        total_failed += 1

    # ------------------------------------------------------------------
    # 8. BarraCUDA connection documented
    # ------------------------------------------------------------------
    print("\n--- Check 8: BarraCUDA connection ---")
    print("  PhyloNet-HMM (Paper 018):")
    print("    States = ILS_only, Introgression")
    print("    Observations = gene tree topologies (concordant, introg-like, other)")
    print("  BarraCUDA mapping: same as Paper 016 (gemm_f64, reduce_max)")
    print("  [PASS] BarraCUDA connection documented")
    total_passed += 1

    total = total_passed + total_failed
    print(f"\n{'=' * 72}")
    print(f"TOTAL: {total_passed}/{total} PASS, {total_failed}/{total} FAIL")
    print(f"{'=' * 72}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
